server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -144,7 +200,175 @@ protected SearchPhase getNextPhase() {
+            return refCounted.tryIncRef();
+        }
+
+        @Override
+        public boolean hasReferences() {
+            return refCounted.hasReferences();
+        }
+
+        @Override
+        public boolean decRef() {
+            if (refCounted.decRef()) {
+                for (int i = 0; i < results.length; i++) {
+                    if (results[i] instanceof RefCounted r) {
+                        r.decRef();
+                    }
+                    results[i] = null;
+                }
+                return true;
+            }
+            return false;
+        }
+    }
+
+    public static class NodeQueryRequest extends TransportRequest implements IndicesRequest {
+        private final List<ShardToQuery> shards;
+        private final SearchRequest searchRequest;
+        private final Map<String, AliasFilter> aliasFilters;
+        private final int totalShards;
+        private final long absoluteStartMillis;
+        private final String localClusterAlias;
+
+        private NodeQueryRequest(SearchRequest searchRequest, int totalShards, long absoluteStartMillis, String localClusterAlias) {
+            this.shards = new ArrayList<>();
+            this.searchRequest = searchRequest;
+            this.aliasFilters = new HashMap<>();
+            this.totalShards = totalShards;
+            this.absoluteStartMillis = absoluteStartMillis;
+            this.localClusterAlias = localClusterAlias;
+        }
+
+        private NodeQueryRequest(StreamInput in) throws IOException {
+            super(in);
+            this.shards = in.readCollectionAsImmutableList(ShardToQuery::readFrom);
+            this.searchRequest = new SearchRequest(in);
+            this.aliasFilters = in.readImmutableMap(AliasFilter::readFrom);
+            this.totalShards = in.readVInt();
+            this.absoluteStartMillis = in.readLong();
+            this.localClusterAlias = in.readOptionalString();
+        }
+
+        @Override
+        public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
+            return new SearchShardTask(id, type, action, "NodeQueryRequest", parentTaskId, headers);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeCollection(shards);
+            searchRequest.writeTo(out, true);
+            out.writeMap(aliasFilters, (o, v) -> v.writeTo(o));
+            out.writeVInt(totalShards);
+            out.writeLong(absoluteStartMillis);
+            out.writeOptionalString(localClusterAlias);
+        }
+
+        @Override
+        public String[] indices() {
+            return shards.stream().map(s -> s.originalIndices().indices()).flatMap(Arrays::stream).distinct().toArray(String[]::new);
+        }
+
+        @Override
+        public IndicesOptions indicesOptions() {
+            return shards.getFirst().originalIndices.indicesOptions();
```

Comment:
You are getting the first one because they are all the same, given they come from the original search request? I wonder if it makes sense to carry around original indices in ShardToQuery, as opposed to only the index name? We can take the indices options from the search request, and there is no need to carry indices in an array?

Reply:
No it doesn't :) I actually found that last week in my branch that has more changes https://github.com/elastic/elasticsearch/compare/main...original-brownbear:elasticsearch:batched-exec-short-and-very-fast :D I just didn't wanna change things here while you were reviewing. Happy to update this here already :)
Note that we still have `org.elasticsearch.action.search.AbstractSearchAsyncAction#getOriginalIndices` to support so it's not a 100% trivial change to make full use of this optimization but we can save a lot of serialization/deserialization here indeed. Give me a couple minutes to fold that in here :)

Reply:
Done folded this in now to only track the array per shard. We could probably keep the original array and build things out from that, but it does make the code trickier I think because we currently build the per-shard arrays when we build the iterators. Maybe worth a follow-up though? (I agree that we obviously send redundant information still, just tricky to fix it without adding logic)